src/cluster/placement/algo/mirrored_test.go

```diff
@@ -646,37 +429,318 @@ func TestMirrorRemoveAndRevertAfterCutover(t *testing.T) {
+	assert.NoError(t, placement.Validate(p2))
+
+	i4, ok = p2.Instance("i4")
+	assert.True(t, ok)
+	assert.True(t, i4.IsInitializing())
+	i5, ok = p2.Instance("i5")
+	assert.True(t, ok)
+	assert.True(t, i5.IsLeaving())
+
+	assert.True(t, ssI4.Equals(i5.Shards()))
+	// Can't directly compare shards.Equals because the shards in ssI5 will be having "i4"
+	// as sourceID and shards in i4 will be having "i5" as sourceID.
+	assert.Equal(t, ssI5.NumShardsForState(shard.Initializing), i4.Shards().NumShardsForState(shard.Initializing))
+}
+
+func TestMirrorMultipleNonOverlappingReplaces(t *testing.T) {
+	i1 := newTestInstance("i1").SetShardSetID(1)
+	i2 := newTestInstance("i2").SetShardSetID(1)
+	i3 := newTestInstance("i3").SetShardSetID(2)
+	i4 := newTestInstance("i4").SetShardSetID(2)
+	instances := []placement.Instance{i1, i2, i3, i4}
+
+	numShards := 100
+	ids := make([]uint32, numShards)
+	for i := 0; i < len(ids); i++ {
+		ids[i] = uint32(i)
+	}
+
+	now := time.Now()
+	nowNanos := now.UnixNano()
+	shardCutoverTime := now.Add(time.Hour).UnixNano()
+	a := NewAlgorithm(placement.NewOptions().
+		SetIsMirrored(true).
+		SetShardCutoverNanosFn(func() int64 { return shardCutoverTime }).
+		SetShardCutoffNanosFn(func() int64 { return shardCutoverTime }))
+	p, err := a.InitialPlacement(instances, ids, 2)
+	assert.NoError(t, err)
+	assert.NoError(t, placement.Validate(p))
+	assert.Equal(t, uint32(2), p.MaxShardSetID())
+
+	p, _, err = a.MarkAllShardsAvailable(p)
+	assert.NoError(t, err)
+
+	// First replace: i1 replaced by i5.
+	i5 := newTestInstance("i5").SetShardSetID(1)
+	p1, err := a.ReplaceInstances(p, []string{"i1"}, []placement.Instance{i5})
+	require.NoError(t, err)
+	require.NoError(t, placement.Validate(p1))
+	assert.Equal(t, uint32(2), p1.MaxShardSetID())
+	assert.True(t, allLeaving(p1, []placement.Instance{i1}, nowNanos))
+	assert.True(t, allInitializing(p1, []string{"i5"}, nowNanos))
+
+	// Second replace that does not overlap with first replace: i3 replaced by i6.
+	i6 := newTestInstance("i6").SetShardSetID(2)
+	p2, err := a.ReplaceInstances(p1, []string{"i3"}, []placement.Instance{i6})
+	require.NoError(t, err)
+	require.NoError(t, placement.Validate(p2))
+
+	assert.Equal(t, uint32(2), p2.MaxShardSetID())
+	assert.True(t, allLeaving(p2, []placement.Instance{i3, i1}, nowNanos))
+	assert.True(t, allInitializing(p2, []string{"i6", "i5"}, nowNanos))
+}
+
+func TestMirrorReplacesCannotOverlap(t *testing.T) {
+	i1 := newTestInstance("i1").SetShardSetID(1)
+	i2 := newTestInstance("i2").SetShardSetID(1)
+	i3 := newTestInstance("i3").SetShardSetID(2)
+	i4 := newTestInstance("i4").SetShardSetID(2)
+	instances := []placement.Instance{i1, i2, i3, i4}
+
+	numShards := 100
+	ids := make([]uint32, numShards)
+	for i := 0; i < len(ids); i++ {
+		ids[i] = uint32(i)
+	}
+
+	now := time.Now()
+	nowNanos := now.UnixNano()
+	shardCutoverTime := now.Add(2 * time.Hour).UnixNano()
+	a := NewAlgorithm(placement.NewOptions().
```

Comment:
A lot of common boilerplate still between tests--worth factoring some out again? Easyish targets:

- Setup method to setup your algorithm
- Test constructor for placement with n shards. This tbh could probably handle your instance creation if you wanted as well; something like:

```
func newTestPlacement(numInstances int, numShards int) placement.Placement {

}

// if you still want i1, i2 vars, could extract them from the placement by key.
```
Could also just do ```
newTestPlacement(instances []placement.Instance, numShards int)
```
if that's preferable; either way.

Reply:
I'm following the style of existing tests: I agree that is somewhat redundant but it is readable. I don't want to create a huge in this code. I want to keep it very well scoped. This improvement could be done in separate PR.

Reply:
Eh sure, not that big of a deal. I think a lot of the changes could be done pretty well in the same way you did the instance constructor change (and I'm definitely happy to review the "irrelevant churn" there), but I'm fine moving forward without it.